### PR TITLE
Don't permalloc the pkgimgs unnecessarily

### PR DESCRIPTION
**Pull request by gbaraldi**
It seems we were permallocing the memory for the sysimg even though the OS already does that more efficiently.

### CODE REVIEW

**Comment by vchuravy** [comment]
I think we still hit this codepath with `--pkgimages=no` where we still would need to do that?

**Comment by gbaraldi** [comment]
Yeah, annoyingly we don't have that information from there. Do you find it reasonable to check options there?

**Comment by vchuravy** [comment]
@gbaraldi I pushed a fix. Can I ask you to write some tests?

**Comment by gbaraldi** [comment]
So I was thinking about the test. Do we spawn a process with pkgimgs=no and load some package? Maybe a stdlib?

File: src/staticdata.c [hunk 1]
```diff
@@ -3414,7 +3423,7 @@ JL_DLLEXPORT jl_value_t *jl_restore_package_image_from_file(const char *fname, j
         }
     #endif
 
-    jl_value_t* mod = jl_restore_incremental_from_buf(pkgimg_data, &pkgimage, *plen, depmods, complete);
+    jl_value_t* mod = jl_restore_incremental_from_buf(pkgimg_data, &pkgimage, *plen, depmods, complete, true);
```
Comment (vchuravy) [inline src/staticdata.c:3426]:
```suggestion
    jl_value_t* mod = jl_restore_incremental_from_buf(pkgimg_data, &pkgimage, *plen, depmods, complete, false);
```

**Review by vtjnash — APPROVED** [review]

**Review by vtjnash — COMMENTED** [review]
bool/true/false is C++, this should use int/1/0

```
Error in testset precompile:
Error During Test at /cache/build/default-amdci4-3/julialang/julia-master/julia-bdc77d5c19/share/julia/test/precompile.jl:339
  Test threw exception
  Expression: begin
    m = Base._require_from_serialized(Base.PkgId(Foo), cachefile, ocachefile)
    #= /cache/build/default-amdci4-3/julialang/julia-master/julia-bdc77d5c19/share/julia/test/precompile.jl:341 =# @test m isa Module
end
  Error reading package image file.
```

**Comment by gbaraldi** [comment]
Doesn't `stdbool.h` add `bool/true/false`?

**Comment by gbaraldi** [comment]
I mentioned this privately to valentin. But the test failure we see there happens because we specifically load the same package image twice. Though since we are dlopening it, instead of mmaping a different copy, we just get the same thing, that means we have already performed relocations over it. So that test is doing something kinda illegal here. Though we might want to add a check that we've already loaded a sysimg before.

File: test/precompile.jl [hunk 2]
```diff
@@ -346,12 +346,12 @@ precompile_test_harness(false) do dir
     end
     # use _require_from_serialized to ensure that the test fails if
     # the module doesn't reload from the image:
-    @test_warn "@ccallable was already defined for this method name" begin
-        @test_logs (:warn, "Replacing module `$Foo_module`") begin
-            m = Base._require_from_serialized(Base.PkgId(Foo), cachefile, ocachefile)
-            @test isa(m, Module)
-        end
-    end
+    # @test_warn "@ccallable was already defined for this method name" begin
```
Comment (vchuravy) [inline test/precompile.jl:349]: Could we just do this test with `--pkgimages=no`?

**Comment by KristofferC** [comment]
> Do we have a check to see if package was loaded beforehand?

The check is typically at https://github.com/JuliaLang/julia/blob/87f8958a218215df0aaffab59736c30eb6f51d2e/base/loading.jl#L1569 but if you call `require` directly then that is skipped and the existing module is overwritten.

**Comment by vchuravy** [comment]
@nanosoldier `runtests()`

**Review by vchuravy — APPROVED** [review]

**Comment by nanosoldier** [comment]
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/48215#issuecomment-1412672284) has completed - possible new issues were detected.
A full report can be found [here](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/0faf69e_vs_134f3e7/report.html).

**Comment by gbaraldi** [comment]
@vchuravy I was talking with @vtjnash and maybe we want to move the data section back to the .ji files. Two reasons 1, we might want to load the same file twice in order to inspect it before relocations, which `dlopen` doesn't allow. The second thing is that controlling how we `mmap` the files means we can maybe use huge pages in their load to save a nice amount of time, since julia boot time is mostly page faults.

**Comment by vchuravy** [comment]
Sure that sounds reasonable. I won't have time to tackle it though.

**Comment by KristofferC** [comment]
Doesn't make sense really for this to be on the milestone imo.

**Comment by gbaraldi** [comment]
It was on the milestone because it was initially a small fix, but on further discussion it's probably not so simple. So I think we can release 1.9 without it. If the fix isn't too difficult we can probably land it on a 1.9 patch

**Comment by vchuravy** [comment]
From the HPC call: This currently breaks PkgInspector. So the idea is to add an opt-out env variable that users of PkgInspector can set to get the copy back.

**Pull request closed without merging**